common/hyperscaler/rules/ruleset.go

```diff
@@ -0,0 +1,143 @@
+package rules
+
+import "fmt"
+
+type PatternAttribute struct {
+	matchAny bool
+	literal  string
+}
+
+type ValidRule struct {
+	Plan                    PatternAttribute
+	PlatformRegion          PatternAttribute
+	HyperscalerRegion       PatternAttribute
+	Shared                  bool
+	EuAccess                bool
+	PlatformRegionSuffix    bool
+	HyperscalerRegionSuffix bool
+	MatchAnyCount           int
+}
+
+type ValidationErrors struct {
+	ParsingErrors   []error
+	DuplicateErrors []error
+	AmbiguityErrors []error
+}
+
+func (pa *PatternAttribute) Match(value string) bool {
+	if pa.matchAny {
+		return true
+	}
+	return pa.literal == value
+}
+
+func (vr *ValidRule) Match(provisioningAttributes *ProvisioningAttributes) bool {
+	if !vr.Plan.Match(provisioningAttributes.Plan) {
+		return false
+	}
+
+	return vr.matchRegions(provisioningAttributes)
+}
+
+func (vr *ValidRule) matchRegions(provisioningAttributes *ProvisioningAttributes) bool {
+	if !vr.PlatformRegion.Match(provisioningAttributes.PlatformRegion) {
+		return false
+	}
+
+	if !vr.HyperscalerRegion.Match(provisioningAttributes.HyperscalerRegion) {
+		return false
+	}
+	return true
+}
+
+func (r *ValidRule) toResult(provisioningAttributes *ProvisioningAttributes) Result {
+	hyperscalerType := provisioningAttributes.Hyperscaler
+	if r.PlatformRegionSuffix {
+		hyperscalerType += "_" + provisioningAttributes.PlatformRegion
+	}
+	if r.HyperscalerRegionSuffix {
+		hyperscalerType += "_" + provisioningAttributes.HyperscalerRegion
+	}
+
+	return Result{
```

Comment:
one note, which is not directly about the PR but can be easy implemented here. Let's add new field to the Result, which contains the rule as a string. So the step (ResolveSubscriptionSecret) calls MatchProvisioningAttributesWithValidRuleset gets not only 3 values used to create label selector. It gets also the original rule just for logging it. At the end of the day, the operator can look at KEB logs and find exaclty which rule was taken into account.

Reply:
Pushed.